Add tests for strided layout in factory functions

alhridoy opened the pull request:
Fixes #111222
This pull request adds tests for factory functions that create tensors with a strided layout. The tests are added to the `test_ops.py` file and check the behavior of the `empty`, `zeros`, `ones`, and `rand` factory functions when used with the `layout=torch.strided` argument.

alhridoy committed:
Add tests for strided layout in factory functions

akdong91 commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,22 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
         self._test_fake_crossref_helper(device, dtype, op, torch.cuda.amp.autocast)
-
+    def test_empty_strided(self, device):
+        result = torch.empty((3, 3), layout=torch.strided, device=device)
```
@alhridoy  Because the default layout for the`empty` and other factory functions is `torch.strided`, is it necessary to specify `layout=torch.strided` explicitly here? Without the layout argument to `empty`, the layout of the result should be `torch.strided`.  I think that needs to be tested here. Let me know if I am missing anything here.

alhridoy replied:
Hi @akdong91 , Thanks for the feedback. I've updated the tests based on your feedback. I've added a generic test that checks all factory functions with layout=torch.strided specified. This test also checks that the default layout is torch.strided when the layout argument is not provided. Please take a look and let me know if there are any further changes needed.

lezcano reviewed (changes requested):
The point of OpInfos is that they give you many inputs on which to test your function. Your task here is to write a test that is run on all the OpInfos that are a factory functions (you can filter these easily) and add `layout=torch.strided`  to their input, and make sure that they indeed work as expected.

All these tests should be generic over the operations, so you should just write one test that will be instantiated for every op

alhridoy committed:
Added test_strided_layout method to test_ops.py

alhridoy commented:
> The point of OpInfos is that they give you many inputs on which to test your function. Your task here is to write a test that is run on all the OpInfos that are a factory functions (you can filter these easily) and add `layout=torch.strided` to their input, and make sure that they indeed work as expected.
> 
> All these tests should be generic over the operations, so you should just write one test that will be instantiated for every op

Hi @lezcano , Thanks for the feedback.  I've updated the tests based on your feedback. I've added a generic test that checks all factory functions with `layout=torch.strided` specified. This test also checks that the default layout is `torch.strided` when the `layout` argument is not provided. Please take a look and let me know if there are any further changes needed.

lezcano commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,15 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
+        for sample in samples:
+            default_result = op(sample.input, *sample.args, **sample.kwargs)
+            self.assertEqual(default_result.layout, torch.strided)
+            strided_result = op(sample.input.to(layout=torch.strided), *sample.args, **sample.kwargs)
```
This is not what yo want to test. You want to pass the layout to the op, not to the function `to`.

alhridoy replied:
@lezcano Thanks for the feedback. Is this what you had in mind? I would appreciate it if you could review these changes.

```
@ops([op for op in op_db if op.is_factory_method], allowed_dtypes=OpDTypes.all)
def test_strided_layout(self, device, dtype, op):
    samples = op.sample_inputs(device, dtype)
    for sample in samples:
        result = op(sample.input, *sample.args, layout=torch.strided, **sample.kwargs)
        self.assertEqual(result.layout, torch.strided)
```

lezcano replied:
no need to pass allowed_dtypes. Also, this would not work if `sample.kwargs` already has `layout` in it, so yeah, almost like this, but not quite

alhridoy replied:
@lezcano I updated the test to be generic over all factory ops, pass layout to the op call directly, and handle kwargs with layout per your feedback. Please let me know if it fits well now.

lezcano commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,15 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
     def test_fake_crossref_backward_amp(self, device, dtype, op):
         self._test_fake_crossref_helper(device, dtype, op, torch.cuda.amp.autocast)
-
+    @ops([op for op in op_db if op.is_factory_method], allowed_dtypes=(torch.float,))
```
Add an empty line before

lezcano replied:
Also, why just floats?

lezcano commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,15 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
+            default_result = op(sample.input, *sample.args, **sample.kwargs)
+            self.assertEqual(default_result.layout, torch.strided)
+            strided_result = op(sample.input.to(layout=torch.strided), *sample.args, **sample.kwargs)
+            self.assertEqual(strided_result, default_result)
```
You don't want to test this as this is not testing the layout. You want to test that the layout is correct

alhridoy committed:
Update test_strided_layout to handle existing layout in kwargs

lezcano commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,15 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
+    def test_strided_layout(self, device, dtype, op):
+        samples = op.sample_inputs(device, dtype)
+        for sample in samples:
+                kwargs = sample.kwargs.copy()
```
This has too many spaces...

lezcano commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,15 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
     def test_fake_crossref_backward_amp(self, device, dtype, op):
         self._test_fake_crossref_helper(device, dtype, op, torch.cuda.amp.autocast)
-
+    @ops([op for op in op_db if op.is_factory_method])
```
Leave an empty line between tests

akdong91 replied:
Shouldn't 'op.is_factory_method` be `op.is_factory_function`?

alhridoy committed:
Added empty line between test methods for readability and fixed spaces

akdong91 commented on test/test_ops.py:
```diff
@@ -2195,7 +2195,16 @@ def test_fake_crossref_backward_no_amp(self, device, dtype, op):
+        samples = op.sample_inputs(device, dtype)
+        for sample in samples:
+            kwargs = sample.kwargs.copy()
+            kwargs['layout'] = torch.strided  
```
With or without this assignment, the test will pass. all factory functions being tested are with the default layout=`torch.strided`.

if i understand correctly, the scope of this test is to test if the result tensor has the layout being specified.
The specified layout can be `torch.strided` or something else.  Doesn't we need to cover the second case to be safe?

lezcano replied:
See the PR linked in the original issue. The test could be more general, but that's out of the scope of this PR

alhridoy replied:
@lezcano @akdong91 Something like this? 
@ops([op for op in op_db if op.is_factory_function])  
def test_tensor_layouts(self, device, dtype, op):
    layouts = [torch.strided, torch.sparse_coo] 
    samples = op.sample_inputs(device, dtype)

    for sample in samples:
        for layout in layouts:
            kwargs = sample.kwargs.copy()
            kwargs['layout'] = layout
            result = op(sample.input, *sample.args, **kwargs)
            self.assertEqual(result.layout, layout)

lezcano replied:
Not for this PR.

akdong91 replied:
> See the PR linked in the original issue. The test could be more general, but that's out of the scope of this PR

That is fair enough.

alhridoy replied:
@akdong91  @lezcano  i beleiev that the following commit Only tests torch.strided layout. Modifies kwargs copy rather than original and Checks result layout matches strided layout kwarg.

alhridoy committed:
Added tests for strided layout in factory functions

lezcano commented:
@pytorchbot merge
